pennylane/ops/op_math/sum.py

```diff
@@ -0,0 +1,156 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the implementation of the Sum class which contains logic for
+computing the sum of operations.
+"""
+import numpy as np
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Operator, expand_matrix, MatrixUndefinedError
+
+
+def sum(*summands, do_queue=True, id=None):
+    """Top level sum function to create an instance of Sum"""
+    return Sum(*summands, do_queue=do_queue, id=id)
+
+
+def _sum(mats_gen, dtype=None, cast_like=None):
+    """Private sum method given a series of matrices of correct size.
+    Super inefficient method just as a proof of concept."""
+
+    res = None
+    try:
+        for i, mat in enumerate(mats_gen):
+            res = mat if i == 0 else math.add(res, mat)
+    except MatrixUndefinedError as error:
+        print(f"\nThe matrix method must be defined for all summands: \n")
+        raise error
+
+    if dtype is not None:
+        res = math.cast(res, dtype)
+    if cast_like is not None:
+        res = math.cast_like(res, cast_like)
+
+    return res
+
+
+class Sum(Operator):
+    """Arithmetic operator subclass representing the sum of operators"""
+
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+
+    def __init__(self, *summands, do_queue=True, id=None):
+
+        if len(summands) < 2:
+            raise ValueError(f"Require at least two operators to sum; got {len(summands)}")
+
+        self.summands = summands
+
+        combined_wires = qml.wires.Wires.all_wires([s.wires for s in summands])
+        combined_params = []
+        for s in summands:
+            combined_params += s.parameters
+
+        super().__init__(*combined_params, wires=combined_wires, do_queue=do_queue, id=id)
+        self._name = "Sum"
+
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        return " + ".join([f"{f}" for f in self.summands])
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.data = self.data.copy()  # copies the combined parameters
+        copied_op.summands = tuple(s.__copy__() for s in self.summands)
+
+        for attr, value in vars(self).items():
+            if attr not in {"data", "summands"}:
+                setattr(copied_op, attr, value)
+
+        return copied_op
+
+    @property
+    def num_wires(self):
+        return len(self.wires)
+
+    @property
+    def is_hermitian(self):
+        """If all of the terms in the sum are hermitian, then the Sum is hermitian."""
+        return all([s.is_hermitian for s in self.summands])
```

Comment:
```suggestion
        return all(s.is_hermitian for s in self.summands)
```
This is a little less cluttered and can be more performant since we don't create the entire array.  We just have to iterate through the generator until we find the first `False`.